File: csharp/Value.cs
using System;

namespace fizzbuzz {
    interface IValue {
        object Data();
    }

    class IntValue : IValue {
        private int data;

        public IntValue(int data) {
            this.data = data;
        }

        public object Data() {
            return data;
        }

        public override string ToString() {
            return data + "";
        }
    }
}

File: haskell/src/Value.hs
module Value(IValue(..), intValue) where

data IValue a = IValue
    { value     :: a
    , toStr     :: String
    , next      :: (IValue a) -> (IValue a)
    , equal     :: (IValue a) -> (IValue a) -> Bool }

intValue :: Int -> (IValue Int)
intValue val =
    IValue  { value =   val
            , toStr =   show val
            , next =    \x -> intValue ((value x) + 1)
            , equal =   \x -> (\y -> (value x) == (value y)) }
